### PR TITLE
[IMP] web_editor, website: add background shapes snippet option

### PR DESCRIPTION
**Pull request by sdegueldre**
This commit adds a new snippet option allowing the user to choose one of
a series of SVG shapes as their background, or to be overlaid over the
background image of a snippet.

this commit also introduces some utility functions to deal with multiple
css backgrounds.

task-2210790

### CODE REVIEW

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 1]
```diff
@@ -3223,6 +3228,143 @@ registry.background = SnippetOptionWidget.extend({
...
+     * @param {string} color the color to normalize into a css value
+     */
+    _normalizeColor(color) {
+        if (!ColorpickerWidget.isCSSColor(color)) {
+            const style = window.getComputedStyle(document.documentElement);
+            color = style.getPropertyValue('--' + color).trim();
+            color = ColorpickerWidget.normalizeCSSColor(color);
+        }
+        return color;
+    },
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:3365]: We are multiplying the location where we do that. Could you do a dedicated commit to share that code somewhere?

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 2]
```diff
@@ -3293,11 +3438,8 @@ registry.BackgroundPosition = SnippetOptionWidget.extend({
      * @override
      */
     selectStyle: function (previewMode, widgetValue, params) {
-        if (params.cssProperty === 'background-size'
-                && !this.$target.hasClass('o_bg_img_opt_repeat')) {
-            // Disable the option when the image is in cover mode, otherwise
-            // the background-size: auto style may be forced.
-            return;
+        if (params.cssProperty === 'background-size') {
+            return setBgImage(this.$target[0], {'background-size': widgetValue});
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:3442]: Seems like this should be part of the base `selectStyle` method instead

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 3]
```diff
@@ -3310,14 +3452,20 @@ registry.BackgroundPosition = SnippetOptionWidget.extend({
...
-        if (methodName === 'backgroundType') {
-            return this.$target.css('background-repeat') === 'repeat' ? 'repeat-pattern' : 'cover';
+        const bgImage = getBgImage(this.$target[0]);
+        switch (methodName) {
+            case 'backgroundType':
+                return (bgImage && bgImage['background-repeat'] === 'repeat') ? 'repeat-pattern' : 'cover';
+            case 'selectStyle':
+                if (params.cssProperty === 'background-size') {
+                    return bgImage ? bgImage['background-size'] : '';
+                }
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:3468]: Same, probably to be part of the base `selectStyle` method

File: addons/website/static/src/js/editor/snippets.options.js [hunk 4]
```diff
@@ -8,6 +8,7 @@ var Dialog = require('web.Dialog');
 const weUtils = require('web_editor.utils');
 var options = require('web_editor.snippets.options');
 const wUtils = require('website.utils');
+const {splitCSSProp, setBackground} = options;
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:11]: Seems useless, this is the only modif in that file

File: addons/website/views/snippets/snippets.xml [hunk 5]
```diff
@@ -294,6 +294,23 @@
         t-att-data-exclude="bg_exclude">
         <t t-call="web_editor.image_optimization_controls"/>
     </div>
+    <div data-js="BackgroundShape" string="Shape"
+        data-selector="section, .parallax, .carousel-item"
+        data-exclude=".s_hr, .s_image_gallery, .o_gallery .carousel-item">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml:299]: Why not using the pretty variable above?

**Comment by sbuhl** [comment]
> Seems to be a nice work 👍 Of course still not convinced about this not being dynamic etc but I guess this is a good first step.
> 
> I have not tested yet, I will once you have reviewed everything following my comments. Also what about the shapes themselves @sbuhl ?

I'll ask a designer (again).

File: addons/web_editor/controllers/main.py [hunk 6]
```diff
@@ -491,3 +493,43 @@ def modify_image(self, attachment, res_model=None, res_id=None, name=None, data=
...
+            'datas': base64.b64encode(svg.encode('utf-8')),
+            'res_model': res_model or 'ir.ui.view',
+            'public': True,
+        }
+        if fields['res_model'] == 'ir.ui.view':
+            fields['res_id'] = 0
+        elif res_id:
+            fields['res_id'] = res_id
+        attachment = original.copy(fields)
+        attachment.url = '/website/background_shape/%s/%s' % (shape, str(attachment.id));
```
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/web_editor/controllers/main.py:533]: Yes, this is used to distinguish svg background-images from background shapes.

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/controllers/main.py:533]: Why do you need both `data-` attributes and a custom URL to distinguish that?

File: addons/web_editor/controllers/main.py [hunk 7]
```diff
@@ -491,3 +493,43 @@ def modify_image(self, attachment, res_model=None, res_id=None, name=None, data=
...
+        fields = {
+            'original_id': original.id,
+            'datas': base64.b64encode(svg.encode('utf-8')),
+            'res_model': res_model or 'ir.ui.view',
+            'public': True,
+        }
+        if fields['res_model'] == 'ir.ui.view':
+            fields['res_id'] = 0
+        elif res_id:
+            fields['res_id'] = res_id
```
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/web_editor/controllers/main.py:531]: How would I do that in the `fields` declaration?

File: addons/web_editor/controllers/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/web_editor/controllers/main.py:533]: I used to parse the shape from the url but it turned out to be a problem for other reasons, so I guess this is no longer needed, good point.

File: addons/web_editor/static/src/js/common/utils.js [hunk 8]
```diff
@@ -167,6 +167,179 @@ function _computeColorClasses(colorNames, prefix = 'bg-') {
     }
     return classes;
 }
+/**
+ * Splits a CSS property into its respective parts, doesn't split between
+ * parentheses. eg:
+ *    splitCSSProp("linear-gradient(to bottom, rgba(255,255,0,0.5), rgba(0,0,255,0.5)), url('/web/static/image.png')", ',')
+ *    => ["linear-gradient(to bottom, rgba(255,255,0,0.5), rgba(0,0,255,0.5))", "url('/web/static/image.png')"]
+ *
+ * @param {string} cssProp the css property to split
```
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/web_editor/static/src/js/common/utils.js:176]: It's especially confusing since you're using cssProp in this very same file, I'll rename it.

File: addons/web_editor/static/src/js/common/utils.js [hunk 9]
```diff
@@ -167,6 +167,179 @@ function _computeColorClasses(colorNames, prefix = 'bg-') {
...
+    //     ...
+    // ]
+    // For why we use padRepeat, see https://www.w3.org/TR/css-backgrounds-3/#layering
+    const splitProps = cssProps.map(([propName, propValue]) => [propName, padRepeat(splitCSSProp(propValue), nbBackgrounds)]);
+    // entriesArray: [
+    //     [["background-image", "url(...)"], ["background-repeat", "repeat"], ...],
+    //     [["background-image", "linear-gradient(...)"], ["background-repeat","no-repeat"], ...],
+    // ]
+    const entriesArray = _.zip(...splitProps.map(([propName, subValues]) => subValues.map(subValue => [propName, subValue])));
+    return entriesArray.map(entries => Object.fromEntries(entries));
```
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/web_editor/static/src/js/common/utils.js:273]: Sure thing. I tried to make the code as clear as possible but this is a lot of shuffling data around so an example is worth a thousand words.

File: addons/website/views/snippets/snippets.xml (hunk 5 above)
Comment (sdegueldre) [reply addons/website/views/snippets/snippets.xml:299]: it didn't exist at the time, fixed

File: addons/web_editor/controllers/main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/controllers/main.py:531]: Is not it already None in any case? So copying to None won't matter? But ok, not important

[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/web_editor/controllers/main.py:531]: None will overwrite things with dict.update so I assumed that passing it to copy would behave the same (how else would you set something to NULL?) but I will double check

**Comment by sdegueldre** [comment]
@qsm-odoo Alright, this should be ready for a first review. Unfortunately the commit history isn't as clean as I would have liked, I messed up and now have a bunch of work on top of the new background-toggler commit which would be hard to untangle to reorder the commits. The main things worth looking at are the final state, and the "remove multiple backgrounds handling" commit in case you want to extract some of the functionality.

The new background switcher is a rough draft since we should probably clarify the how it should work before anything else and there's a few minor technical annoyances, but the basic behaviour is there. I'll also need to change the imagepicker UI to display video url/image name, and find a way to have the video imagepicker appear below all the image options.

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 10]
```diff
@@ -3319,6 +3417,271 @@ registry.BackgroundImage = SnippetOptionWidget.extend({
...
+        const json = this.$target[0].dataset.oeShapeData;
+        return json ? Object.assign(defaultData, JSON.parse(json)) : defaultData;
+    },
+    /**
+     * Returns the default colors for the currently selected shape.
+     *
+     * @private
+     */
+    _getDefaultColors() {
+        const $shapeContainer = this.$target.find('>.o_we_shape')
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:3668]: `> .` (not the first I saw, this one was the one too many who forced me to tell you ^^)

File: addons/web_editor/static/src/scss/web_editor.frontend.scss [hunk 11]
```diff
@@ -31,3 +31,21 @@
...
+    transform: scale(-1,1);
+}
+
+.o_we_flip_y {
+    transform: scale(1,-1);
+}
+
+.o_we_flip_x.o_we_flip_y {
+    transform: scale(-1,-1);
+}
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/scss/web_editor.frontend.scss:51]: No blank lines between related css rules + add a space after `,` inside `scale` (or actually use `scaleX(-1)`, `scaleY(-1)` and `scale(-1)` which would be more explicit)

File: addons/web_editor/static/src/scss/web_editor.frontend.scss [hunk 12]
```diff
@@ -31,3 +31,21 @@
         }
     }
 }
+
+.o_we_shape {
+    @include o-position-absolute(0,0,0,0);
+    position: absolute !important;
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/scss/web_editor.frontend.scss:37]: Why?

File: addons/web_editor/static/src/scss/wysiwyg_snippets.scss [hunk 13]
```diff
@@ -443,7 +443,7 @@ body.editor_enable.editor_has_snippets {
                     margin-bottom: 1px; // Not sure why but not really centered otherwise
                 }
             }
-            &.active {
+            &.active:not([data-shape]) {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/scss/wysiwyg_snippets.scss:446]: Don't make this something specific to shapes but rather to a class you can add on any select/button

File: addons/web_editor/static/src/scss/wysiwyg_snippets.scss [hunk 14]
```diff
@@ -1435,3 +1448,25 @@ body.editor_enable.editor_has_snippets {
...
 }
+
+we-select.o_we_shape_menu {
+    position: static !important;
+
+    we-button[data-shape] {
+        padding: 0 !important;
+
+        &.active {
+            border: 1px solid var(--success) !important;
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/scss/wysiwyg_snippets.scss:1459]: Use `$o-we-success`

File: addons/web_editor/static/src/scss/wysiwyg_snippets.scss [hunk 15]
```diff
@@ -1435,3 +1448,25 @@ body.editor_enable.editor_has_snippets {
...
+        }
+
+        & > div {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/scss/wysiwyg_snippets.scss:1462]: Useless `& >`

File: addons/web_editor/static/src/scss/wysiwyg_snippets.scss [hunk 16]
```diff
@@ -1435,3 +1448,25 @@ body.editor_enable.editor_has_snippets {
...
+            width: 100%;
+        }
+
+        .o_we_shape_btn_content {
+            width: 100%;
+            height: 75px;
+            background-image: url('/web/static/src/img/transparent.png');
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/scss/wysiwyg_snippets.scss:1469]: Is the related mixin to add this transparent image not working here?

File: addons/web_editor/views/snippets.xml [hunk 17]
```diff
@@ -105,6 +102,62 @@
...
+                    <we-button data-shape="website/shape_bg_2">Shape 2</we-button>
+                    <we-button data-shape="website/shape_bg_3">Shape 3</we-button>
+                    <we-button data-shape="website/shape_bg_4">Shape 4</we-button>
+                    <we-button data-shape="website/shape_bg_5">Shape 5</we-button>
+                    <we-button data-shape="website/shape_bg_6">Shape 6</we-button>
+                    <!-- Would be better to have the x/y flip as a separate data-attribute, but this is
+                    not possible with a current system, as the method is triggred on the select, and method
+                    params are unique per data-method on a given widget -->
+                    <we-button data-shape="website/shape_bg_1,x">Shape 1 flip-x</we-button>
+                    <we-button data-shape="website/shape_bg_2,xy">Shape 2 flip-xy</we-button>
```
Comment (qsm-odoo) [inline addons/web_editor/views/snippets.xml:124]: If I understand your explanation, this is not true anymore since: https://github.com/odoo/odoo/commit/a30cc3402d1ba2678bf2219a33972aff3ef65592

You may probably simplify your code then

File: addons/website/static/src/scss/website.scss [hunk 18]
```diff
@@ -1012,3 +1012,35 @@ $ribbon-padding: 100px;
...
+
+            // eg: o_website_shape_bg_1
+            &.o_#{$module}_#{$shape} {
+                background-image: url("/web_editor/shape/#{$module}/#{$shape}.svg?#{str-slice($url-params, 2)}");
+                background-position: map-get($style, position);
+                background-size: map-get($style, size);
+            }
+        }
+    }
+}
```
Comment (qsm-odoo) [inline addons/website/static/src/scss/website.scss:1046]: I had in mind those classes were not necessary anymore... but maybe I got confused, we need to talk about this tomorrow

File: addons/web_editor/static/src/scss/web_editor.frontend.scss (hunk 12 above)
Comment (sdegueldre) [reply addons/web_editor/static/src/scss/web_editor.frontend.scss:37]: because the background video class will set the position of all children in a section to relative and is more specific

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 19]
```diff
@@ -1579,6 +1583,87 @@ const RangeUserValueWidget = UserValueWidget.extend({
...
+
+        const pagerControls = document.createElement('div');
+        pagerControls.classList.add('d-flex', 'mb-1', 'align-items-center', 'justify-content-end');
+        pagerControls.appendChild(prev);
+        pagerControls.appendChild(this.pageNum);
+        pagerControls.appendChild(next);
+
+        await _super(...arguments);
+        this.menuEl.classList.add('p-2');
+        this.menuEl.prepend(pagerControls);
```
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1618]: > shapes use 3 colors but only 2 colorpickers

That should not be the case. Some colors in the shapes are static colors and can't be customized. There is a single 3-color shape in this branch and it does have 3 color pickers.

File: addons/web_editor/views/snippets.xml (hunk 17 above)
Comment (sdegueldre) [reply addons/web_editor/views/snippets.xml:124]: I would need to extend `widget.setValue` to pass params as well so that I can show the correct shape as active, would that be fine?

File: addons/web_editor/static/src/scss/web_editor.frontend.scss (hunk 12 above)
Comment (qsm-odoo) [reply addons/web_editor/static/src/scss/web_editor.frontend.scss:37]: Meh. To improve later but at least explain that as a comment for now

**Comment by qsm-odoo** [comment]
> Some colors in the shapes are static colors and can't be customized
@sbuhl really? :(

File: addons/web_editor/views/snippets.xml (hunk 17 above)
Comment (qsm-odoo) [reply addons/web_editor/views/snippets.xml:124]: Pfft :/ No... In fact I am not sure to understand this feature. Do you have shapes which are already flipped in the shape selector? Then if you chose them and flip them... would that be possible to distinguish from the unfliped version? Is this really needed?

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 20]
```diff
@@ -3242,6 +3299,49 @@ registry.BackgroundImage = SnippetOptionWidget.extend({
...
+     */
+    toggleBgImage(previewMode, widgetValue, params) {
+        if (this._computeWidgetState('background')) {
+            return this.background(false, '', {});
+        } else {
+            const [bgImgPicker] = this._requestUserValueWidgets('bg_image_opt');
+            if (bgImgPicker) {
+                this.trigger_up('snippet_edition_request', {exec: () => bgImgPicker.el.click()});
+            }
+        }
```
[the remark this answers is not in the document]
Comment (sdegueldre) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3315]: I could try to improve the trigger system, as it is very bare bones and quite hacky right now, I'm just not sure if doing it in this task is a great idea. It would probably be pretty involved. I'm thinking that widgets should only set their own values and trigger themselves through common methods on the `UserValueWidget` base class, this way both of those methods are common to all userValueWidgets and they have a common interface that's public and accessible to other widgets, then we can have a `data-trigger` that would do the same thing as a user-interaction (toggle a checkbox, open a select, open a media-dialog, ...), and a `data-set-widget-value="widgetName: widgetValue, otherWidget: otherValue"` that would do the same thing as selecting a specific value (eg setting a checkbox to being checked, picking a specific option in a select, selecting a specific image in the media-dialog, ...)

This sounds like it should work and be more flexible, but it would require rewriting a few parts of the current userValueWidgets. From what I understand we're trying to get this task merged ASAP so that the designers can work on shapes and integrate them into thems and stuff, so maybe this should better be done separately at a later point? While the "click on the imagepicker el" trick does sound very hacky, it's pretty much what the system I'm describing would achieve except without dispatching an event, so if we intend to rework the trigger system soon I think this hack is acceptable.

[the remark this answers is not in the document]
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3315]: Ok, but at least consider my first two points then

File: addons/web_editor/views/snippets.xml (hunk 17 above)
Comment (sdegueldre) [reply addons/web_editor/views/snippets.xml:124]: Might need to talk about this with SBU/designers. There are a few shapes that are the same except they're flipped, but maybe it's ok to duplicate the svg files for these so we don't have this headache.

Currently, since they're the same shape, if a shape is flipped, I try to find a corresponding button in the select with that x/y flip, and if there is none, I set the unflipped version as active. So if you flip the `Shape 1 flip-x`, it will show `Shape 1` as active, and if you flip `Shape 1` on the x axis, it will show `Shape 1 flip-x` as active, it's not super great but it's not awful either

Comment (qsm-odoo) [reply addons/web_editor/views/snippets.xml:124]: Yes that's what I expected by reading your code and it is confusing to me. The code would be easier without this and the functional behaviour too. Let's duplicate those SVG files.

File: addons/website/static/src/scss/website.scss (hunk 18 above)
Comment (sdegueldre) [reply addons/website/static/src/scss/website.scss:1046]: Leaving a comment for posterity as we discussed this in person.

These classes allow non-customized shapes to update in real-time when changing the theme's color palette, which was a feature that FP really wanted.

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 21]
```diff
@@ -1357,81 +1357,111 @@ const ColorpickerUserValueWidget = SelectUserValueWidget.extend({
...
+        return mediaDialog;
     },
     /**
      * @override
      */
-    getMethodsParams: function (methodName) {
-        return _.extend({isVideo: this.isVideo}, this._super(...arguments));
+    async _updateUI() {
+        this._super(...arguments);
+        this.el.classList.toggle('active', this.isActive());
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:1416]: @sdegueldre I push-forced the branch so don't touch it anymore please :) I reviewed the two first commits in details, I keep the big one for this week-end/monday.

- I added a commit of mine before your work: removing check icon for we-button active state
- I changed the design to not look like a big button, just the fa-edit icon
- I restored the blog option: the active state was lost because you remove the 'active' class addition here
- I changed the hierarchy to add spaces for multiple level hierarchy in options (we'll see with foldable sections later anyway but that way it is already more clear to me)

File: addons/web_editor/views/snippets.xml [hunk 22]
```diff
@@ -113,6 +116,57 @@
...
+
+        <div data-js="BackgroundShape" string="Shape"
+             t-att-data-selector="selector"
+             t-att-data-exclude="exclude"
+             t-att-data-target="target">
+            <we-select-pager class="o_we_shape_menu" string="Shape" data-dependencies="!shape_none_opt" data-name="bg_shape_opt">
+                <!-- Invisible None button to be used for select's placeholder text -->
+                <we-button data-shape="" data-name="shape_none_opt"></we-button>
+                <we-select-page>
+                    <we-button data-shape="website/shape_bg_1" data-select-label="Shape 1"></we-button>
```
Comment (qsm-odoo) [inline addons/web_editor/views/snippets.xml:128]: XML with mention of website inside web_editor... :wink: + the JS option is defined in web_editor but the related css for it to work is defined in website... :wink:

**Comment by sdegueldre** [comment]
>     2. Speaking of future-proof... I really have a hard time with the shape being read from the repo files at visitors' visits. What if we ever want to replace or even move those shapes? Probably best to confirm with AL/FP before the release (kinda related to point 1)

Not sure I understand this.

>     3. Can you lift the mystery around the `str-replace` function you duplicated? :)

I wasn't aware we already had this in the code-base. No other reason, will remove.

>     4. Always use quoted string in SCSS when you can (no problem here but in general: an unquoted string can actually refer to something else. Like `black` which is a color and `'black'` which is a string and both can be distinct keys in a map).

Duly noted

>     5. Here is the rendering of a two snippets with two shapes using the exact same colors:
>        Screenshot from 2020-08-03 16-42-23
>        -> Why can I see the separation between the two snippets?

Because those are shapes created by sbu in a hurry and the gradients don't match, no technical issue here.

>     6. Here is the exact same website as point (5)... but with Chrome:
>        Screenshot from 2020-08-03 16-42-34
>        -> Chrome does not seems to respect your `auto` rule for the y axis `background-size`

Well that's not great :/ I will dig into that. We can probably get around that with the right `preserveAspectRatio` on the svgs but I was hoping that manual SVG editing would not be required. 

>     7. I know the shapes will change but could we avoid to merge shapes containing SBU's trigram? :)

I don't think there was any plans to merge the current shapes, those are just shapes that I've used so sbu can test them and see how the system behaves.

>     8. Speaking of shapes that will change... any news? Again: the feature is really cool... but right now I find _all_ the shapes to be really bad. Second page of shapes is ok but that's because they are really simple. Last page is so bad to me :/

EDI is working on them, they should be ready soon and they look amazing :+1: 

>     9. Maybe part of the ugliness is due to the primary->secondary gradient by default. I wonder if it would not be better with black to white or primary to gray or primary to darker primary ?

As you might have seen in the scss this is all configured per shape, and most of the final shapes will not be primary -> secondary. No worries.

>     10. I asked for this multiple times... but did we in fact let go of the shape selector opening at the right page? In fact since we now have the shape toggler it is even further that than: can we select the "right next shape" when toggling a shape? So if I drop a snippet, add shape "page 2, number 3", then drop a snippet below, add shape... it automatically select "page 2, number 4" by default.

I got caught up in the git history and background-toggler stuff so I forgot. I will look into it but we'll need to clarify functional specs for this part.

>     11. Can we improve the we-select-pager so that the pager is always at the top of the select, outside of the scrollable content (would be better to keep it when the page is scrolled but also: some page have a scrollbar and some do not so it is hard to switch pages.

Indeed. I left-aligned it because that's how it was in the mockup but since the scroll bar alters the layout, you can't just keep your mouse over the next button and keep clicking which is an issue. 

> Maybe lots of points could be discussed with @sbuhl, AL and/or FP.
> 
> You can now push on the branch, just don't forget to fetch it first +1

Sounds good, let's talk about some of the issues you raised tomorrow +1

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 23]
```diff
@@ -3342,31 +3437,289 @@ registry.BackgroundImage = SnippetOptionWidget.extend({
...
+            if (shapeContainer) {
+                shapeContainer.remove();
+            }
+            return;
+        }
+        if (!shapeContainer) {
+            shapeContainer = document.createElement('div');
+            target.prepend(shapeContainer);
+            target.style.position = 'relative';
+            shapeContainer.nextElementSibling.style.position = 'relative';
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:3638]: @sdegueldre Can you explain this to me? I fear this is not needed at all but in any case, this is not robust enough.

File: addons/website/views/snippets/snippets.xml [hunk 24]
```diff
@@ -121,6 +121,27 @@
...
+    <xpath expr="//we-button[@data-toggle-bg-image]" position="after">
+        <we-button title="Video" data-toggle-bg-video="true" data-no-preview="true" data-name="bg_video_toggler_opt">
+            <i class="fa fa-fw fa-film"/>
+        </we-button>
+    </xpath>
+    <xpath expr="//div[@data-js='BackgroundOptimize']" position="after">
+        <div data-js="BackgroundVideo"
+             t-att-data-selector="selector"
+             t-att-data-exclude="exclude"
+             t-att-data-target="target">
```
Comment (qsm-odoo) [inline addons/website/views/snippets/snippets.xml:134]: I think there is a missing `t-att-data-dependencies="images_dependencies"` here (the dependencies that was there for the picker which allowed both image and video)

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 23 above)
Comment (sdegueldre) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3634]: I honestly don't remember. Removing it for now, I think it was to handle some edge case but the code has changed so much since then that it's entirely possible it's no longer needed.

Comment (sdegueldre) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3634]: Same as with parallax, since the shape is absolutely positioned it will show on top of non-positioned elements, but as you said this is not very robust. In parallax we do something like

```
.parallax > * {
    position: relative;
}
```

so I will do something similar instead:

```
[data-oe-shape-data] > * {
    position: relative;
}
```

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:3634]: Ok, I'll merge something better once shape is merged anyway ;)

**Comment by qsm-odoo** [comment]
@sdegueldre I push-forced the branch so don't forget to fetch. What I did:
- Squashed your commits
- Removed the `<folder>` addition in shape route
- Reviewed the scss for the select-pager so there is no `calc` and absolute-positioned stuff
- Fix the web_editor/primary_variables.scss mistake
- Removed the chess-like BG of the shape dropdown and replaced it with a gray that works for all shapes (it is easier that finding the section one as FP wanted but it is also better IMO, if the section is white you cannot see many shapes)

I also added a commit that I will merge aside. I just wanted to check if it worked well with shapes: no more flickering when going from one shape to another :wink: 

Focus on having EDI's shape "refactored" in the repo tomorrow ;) I'll merge whatever we have, thanks ! :3

**Comment by qsm-odoo** [comment]
@sdegueldre I rebased to remove the merged commit then I made this small change https://github.com/odoo/odoo/compare/36636f53ae60e252eafd37b2aec4427c4abedd98..df041cd4b729d8a781b929cb64c8b73e0822b070 to have the same as what you did for image... :) It is actually more clear: show the toggler for video if image_dependencies and show the video picker if there is a video :)

I added some things to fix post-merge in the task pad (the biggest one being the 1px gap between shape with a suggested ugly solution, but it works).

Meanwhile:
@robodoo rebase-ff r+
